core/lib/dal/src/models/consensus/tests.rs

```diff
@@ -0,0 +1,105 @@
+use zksync_protobuf::repr::{decode, encode};
+use zksync_types::{ExecuteTransactionCommon, Transaction};
+
+use crate::tests::{mock_l1_execute, mock_l2_transaction, mock_protocol_upgrade_transaction};
+
+/// Tests struct <-> proto struct conversions.
+#[test]
+fn test_encoding() {
+    encode_decode(mock_l1_execute().into());
+    encode_decode(mock_l2_transaction().into());
+    encode_decode(mock_protocol_upgrade_transaction().into());
+}
+
+fn encode_decode(tx1: Transaction) {
+    let bytes = encode::<super::proto::Transaction>(&tx1);
+    let tx2 = decode::<super::proto::Transaction>(&bytes).unwrap();
+    assert_transaction_eq(&tx1, &tx2);
+}
+
+/// Checks equality. Top-level `PartialEq` of `Transaction` checks only
+/// the hash hence it's not usable here.
+fn assert_transaction_eq(tx1: &Transaction, tx2: &Transaction) {
+    assert_eq!(tx1.received_timestamp_ms, tx2.received_timestamp_ms);
+    assert_eq!(tx1.raw_bytes, tx2.raw_bytes);
+    assert_eq!(tx1.execute, tx2.execute);
+    match (&tx1.common_data, &tx2.common_data) {
```

Comment:
this doesn't look good. Instead define a custom type X, which has all the data from Transaction that you want to compare for exact equality, then implement conversion Transaction -> X, and derive PartialEq for X.

Reply:
Fixed, changes here: https://github.com/matter-labs/zksync-era/pull/1047/commits/d0dc588b8818a7eb9c3143a52d596813ec704a99